### PR TITLE
Cranelift: refactoring of unwind info

### PR DESCRIPTION
**Pull request by yurydelendik**
This PR:
- Creates common unwind info data structures (at `cranelift::isa::unwind::input`)
- Extracts common logic for SystemV and FastCall into "cranelift/codegen/src/isa/x86/unwind.rs"
- Moves emit logic from cranelift/codegen/src/isa/x86/unwind/winx64.rs to cranelift/codegen/src/isa/unwind/winx64.rs
- Moves emit logic from cranelift/codegen/src/isa/x86/unwind/systemv.rs to cranelift/codegen/src/isa/unwind/systemv.rs
- Fixes test at x86/unwind/systemv.rs

This is a preparation PR to expose unwind info for different backends, e.g. new backend for Aarch64. Also, will be useful to move emitting logic out of the codegen crate and keep only common data structures.

See also #2266

### CODE REVIEW

File: cranelift/codegen/src/isa/unwind.rs [hunk 1]
```diff
@@ -14,3 +14,55 @@ pub enum UnwindInfo {
...
+        },
+    }
+
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+    pub struct UnwindInfo<Reg> {
+        pub(crate) prologue_size: CodeOffset,
+        pub(crate) prologue_unwind_codes: Vec<UnwindCode<Reg>>,
+        pub(crate) epilogues_unwind_codes: Vec<Vec<UnwindCode<Reg>>>,
+        pub(crate) code_len: CodeOffset,
```
Comment (peterhuene) [inline cranelift/codegen/src/isa/unwind.rs:66]: nit: perhaps call this `function_size` to correspond with `prologue_size`?

**Review by peterhuene — COMMENTED** [review]
Overall, the refactoring seems pretty straight-forward and most of my comments are simply about the naming of the ABI-agnostic names of the unwind operations.

I do have a concern about the performance of this change, primarily for Windows.  The Windows x64 ABI only needs to look at the entry block to fully describe the unwind information because epilogues are not described for Windows x64.  Epilogues on Windows x64 must have a specific form and Windows will detect an unwind inside of an epilogue and emulate the completion of the epilogue to unwind the function.

With these changes, the more complex scan needed to build DWARF CFI is used even on Windows.  This includes sorting the blocks and then inspecting every instruction in the function.  I would imagine this can be pretty expensive for large modules.  There's also now the additional allocations being done for building up this intermediary "ABI-agnostic" unwind information and then feeding it to the `build` function that transforms it to the "ABI-specific" unwind information.

Obviously these are merely concerns until some actual performance measurements are made, but it's something to think about.

Regarding Windows ARM64, I think this refactoring won't introduce any limitations for emitting unwind information 👍.  Unlike Windows x64, the ARM64 unwind information *does* describe the epilogues, but in the majority case the epilogues are described in terms of the prologue (i.e. there would just be a set of unwind codes for the prologue and the epilogue unwind information would point at the same set of codes to indicate a "mirror" operation).  However, this means the ARM64 unwind information will require the offset to the start of each epilogue within the function.

That said, I haven't looked at what the ARM backend emits for prologues and epilogues and whether or not it will emit noncontiguous functions (complicates the unwind information a little).  If the ARM backend only ever emits "mirror" epilogues (i.e. an epilogue that has a 1:1 reverse instruction for each prologue instruction), then, as with the x64 unwind information, the Windows ARM64 unwind information would ever only need to look at the entry block.

File: cranelift/codegen/src/isa/unwind.rs [hunk 2]
```diff
@@ -14,3 +14,55 @@ pub enum UnwindInfo {
...
+        },
+        StackAlloc {
+            offset: CodeOffset,
+            size: u32,
+        },
+        StackDealloc {
+            offset: CodeOffset,
+            size: u32,
+        },
+        SetCfaRegister {
```
[the remark this answers is not in the document]
Comment (peterhuene) [reply cranelift/codegen/src/isa/unwind.rs:48]: Note that Windows ARM64 unwind information will have a "set_fp" unwind code.

File: cranelift/codegen/src/isa/unwind.rs [hunk 3]
```diff
@@ -14,3 +14,55 @@ pub enum UnwindInfo {
...
+    pub(crate) enum UnwindCode<Reg> {
+        PushRegister {
+            offset: CodeOffset,
+            reg: Reg,
+        },
+        PopRegister {
+            offset: CodeOffset,
+            reg: Reg,
+        },
+        SaveXmm {
```
[the remark this answers is not in the document]
Comment (yurydelendik) [reply cranelift/codegen/src/isa/unwind.rs]: Windows x64 emit logic wants to differentiate these two. I'll rename it to `SaveXmmRegister`, I'll defer this refactoring to different PR -- it is getting more complicated than simple refactoring.

**Review by peterhuene — APPROVED** [review]
Looks great!  Storing the block with the epilogue starts definitely makes sense.

**Pull request merged**